### ISSUE TITLE
DataTable: JS error on return from Dialog

### ISSUE THREAD

**Issue by brooksie** [primefaces/primefaces#6957]
**Describe the defect**
On return from a Dialog (using Dialog Framework) a JS error occurs when refreshing the old page with the new data (created in the dialog).
![consoleError](https://user-images.githubusercontent.com/4806551/107074626-e78c5100-67e0-11eb-8e4f-32ab0ffdd6aa.png)
The list page (that initiates the dialog) is *occasionally* not refreshed on return and gets "stuck".

**Reproducer**
See attached. This error occurs with both 8.0 and 10.0RC1. I haven't tried Snapshot.
[DataTableJSError.zip](https://github.com/primefaces/primefaces/files/5935014/DataTableJSError.zip)

**Environment:**
 - PF Version: 8.0 and 10.0RC1
 - JSF + version: Both Mojarra2.3 and MyFaces 2.3 (reproducer requires at least 2.3)
 - Affected browsers: Firefox and Chrome (others not tested)

**To Reproduce**
1. Notice the data table list on the initial page. Press "Add" to open dialog to add a new row to the list.
2. Enter "EL" (or anything not in the current list) and press "Save".
3. The JS error is shown in the Console. The List is *occasionally* not updated with the new row.
4. To check if the row really was saved, refresh the whole page.

The same issue also shows for the Delete and Edit options. It isn't clear to my why the list is usually updated but sometimes not. The server log file includes a list of the content of the list immediately before it is returned to JSF and it clearly shows the update to the list having been made in all cases.

**Expected behavior**
No errors. The screen is refreshed consistently to show the latest data list

**Example XHTML**
See reproduced.

**Example Bean**
See reproducer.

**Comment by melloware** [comment]
Everything seems to be working fine and that is only logged in JSF Development Stage right?
```javascript
 if (PrimeFaces.isDevelopmentProjectStage() && window.console) {
                console.error(log);
            }
```

**Comment by brooksie** [comment]
Yes and no.The message only appears in Development mode. It's absent in production mode. 
Mostly things are working ok - but just occasionally the screen doesn't refresh as expected on return from the dialog. (The row that was added doesn't appear or the row that was deleted remains displayed, or the screen sort of freezes) As I said in the ticket, I'm not sure why and I'm having a hard time reproducing the problem consistently. The console message was the only indication I could find of a problem, so I logged it on that basis. This all started with issues in filtering the table and particularly the call to PF('...').filter(), hence why I included filter options. Sorry this isn't more concrete.


    On Friday, 5 February 2021, 20:26:39 GMT, Melloware <notifications@github.com> wrote:  
 
 


Everything seems to be working fine and that is only logged in JSF Development Stage right?
 if (PrimeFaces.isDevelopmentProjectStage() && window.console) {
                console.error(log);
            }

**Comment by melloware** [comment]
OK so in this tester I am not getting it to freeze at all so I think that JS error is a red herring its just trying to update the dialog that has already been destroyed.

**Comment by brooksie** [comment]
I'm not sure about the red herring, but I guess that unless you want to keep the ticket open to resolve the JS error message, it will have to be closed as not proven (or whatever category you use) unless I can get some harder evidence of an issue. I understand.

    On Friday, 5 February 2021, 20:43:02 GMT, Melloware <notifications@github.com> wrote:  
 
 


OK so in this tester I am not getting it to freeze at all so I think that JS error is a red herring its just trying to update the dialog that has already been destroyed.

**Comment by melloware** [comment]
well I will leave it open for now hoping you can add some debug and figure out where your scenario is locking up?

**Comment by brooksie** [comment]
OK. This version is more concrete. The filter comments previously could be a red herring It is related to the return from a dialog. Here is the reproducer:
[DataTableUpdateError.zip](https://github.com/primefaces/primefaces/files/5938948/DataTableUpdateError.zip)

The concept is to change the membership of the many-to-many relationship User-Roles. The list page has two options - (i) to see membership by user or (ii) to see it by role. The options to add or delete membership call the same dialog to handle a single row change, then returns back to the list. The steps to reproduce are:
0. Select either version (the only difference being no "oncomplete=...filter()" on the dialogreturn ajax)
1. Click "byUser"
2. Select user "Chris"
3.    Notice that 1 role ("*.view") is listed for Chris.
4. Click "byRole"
5. Select role "User.create"
6.    Notice 2 users (Jo, Mike) are listed for role "User.create". [6A. You should be able to click ByUser and ByRole and swap backwards and forwards between the two views with the table contents and columns changing each time. The server log shows the datamodel changing on each click.]
7. Click "Add" to add a new user to this role
8. Select user "Judith"
9. Click "Add" to create the membership relationship and return to the main list
10.   Notice now that 3 users (Judith, Jo, Mike) are listed for role "User.create"
11. Click "byUser"
12. ERROR - Screen shows 3 rows and all rows say "User.create". The output ought to be identical to step 3 for user Chris with one row for role "*.view".

If in step 11 you don't click byUser but instead change the selected role, it also fails to refresh datatable. It appears to be "frozen".

Steps 11 and 12 show that the screen is being updated successfully because the titles and column displayed is changing. However the data table contents (datamodel) are not being updated. Step 12 shows 3 rows because there are 3 rows in step 10 and the datamodel has not been changed since then . All the values are the same because in step 10 there is a hidden column containing roles which were all identical (it is a by role view). Step 12 is byUser, so the user column is hidden and the role column is shown instead -  BUT the server log clearly shows that the updated data is being returned each time the ByOption or selectedUser or selectedRole is changed.

Step 6A shows the JSF is working ok before the dialog is called. If just starts to fail after the return. 

NB. as before, this fails on both PF8 and PF10RC1. I havent tried Snapshot.  It fails with both Mojarra23 and Myfaces23.
Sorry this is not a simple reproducer, but thank-you for investigating.

**Comment by brooksie** [comment]
Oops... pressed the wrong button.

**Comment by melloware** [comment]
I made two small changes and it seems to work now.
[pf-6957.zip](https://github.com/primefaces/primefaces/files/5939067/pf-6957.zip)

1. On the Add button in the dialog I removed `update="@form``

2. On the dialogReturn I changed `update="@form"` to `update="pnlTable"` and removed the `oncomplete="filter()"`.   Not sure if I needed to remove the filter() or not but definitely changing your update to only update the panel and not the whole form seems to fix it.

**Comment by brooksie** [comment]
Unfortunately I can still make it fail sometimes using a different scenario. For example:
1. Click ByUser
2. Select user "Mike"
3. Use the filter to filter on "User." The list will reduce to 4 rows.
4. Click on ByRole
5. Select "User.create"
6. ERROR - instead of 4 rows, there should be 2, for users Jo and Mike.
So "dialogReturn" cannot be the culprit...

**Comment by melloware** [comment]
I think its all your update="@form"  try and be more precise with your updates.

**Comment by brooksie** [comment]
Changing `update=` didn't make any difference. I think "update" only impacts *which* components get updated, not *how* they are updated.

I have resolved my issue by making separate DataTables for the "ByUser" and "ByRole" views rather than a single table, which has removed the need to have columns which are optionally rendered/visible depending on the view. I also needed to ensure the filteredValue list was nulled each time the underlying datamodel is changed.

I have found how to make the datatable freeze. If a column is filtered (or has been filtered but isn't any more), and then the underlying datamodel is changed by an Ajax listener, the table is not updated on the sceen and remains frozen, even with `update="datatable"` and the filteredValuelist has been cleared. By changing the filter value manually, the table unfreezes to show the new data. This can be resolved by adding `oncomplete="PF('tableWidget').filter();` to the ajax command.
